phc/easy/query/fhir_dsl_query.py

```diff
@@ -0,0 +1,87 @@
+from toolz import pipe, identity, curry
+from typing import List, Union
+from lenses import lens
+
+FHIR_WHERE = lens.Get("where", {})
+FHIR_WHERE_TYPE = FHIR_WHERE.Get("type", "")
+FHIR_SIMPLE_QUERY = FHIR_WHERE.Get("query", {})
+FHIR_BOOL_QUERY = FHIR_SIMPLE_QUERY.Get("bool", {})
+
+
+@curry
+def and_query_clause_terms(second_query_clause, first_query_clause):
+    return {
+        "bool": {
+            "should": [first_query_clause, second_query_clause],
+            "minimum_should_match": 2,
+        }
+    }
+
+
+def and_query_clause(query: dict, query_clause: dict):
```

Comment:
This function uses lenses to immutably update the query passed to it. It makes the logical branches much cleaner especially since I plan to build on this in the future.

Reply:
It will only update the query if it knows how to. Otherwise, if you use an argument with a complicated ES query that it doesn't understand, it will error out. That way, we have a "best effort" approach.